modules/storage/src/StockStorageAPI.php

```diff
@@ -21,180 +21,200 @@ public function createTransaction($variation_id, $location_id, $zone, $quantity,
-
-    // Total all transactions between last and max transactions.
-    // @todo - need to use a higher level method.
-    $query = "SELECT location_id, sum(qty) as transactions_qty FROM `{cs_inventory_transaction}`
-      WHERE (`location_id` = '" . $location_id . "')
-      AND (`variation_id` =" . $variation_id . ")
-      AND (`trid` > " . $last_transaction . ")
-      AND (`trid` <= " . $max_transaction . ")
-      GROUP BY location_id";
-    $result = $db->query($query)->fetch();
-    if ($result) {
-      // Add the transactions qty to the existing location level.
-      $stock_level += $result->transactions_qty;
-      // @todo - use non Deprecated function.
-      $io = db_update('cs_inventory_location_level');
-      $io->fields([
-        'qty' => $stock_level,
-        'last_transaction_id' => $max_transaction,
-      ]);
-      $io->condition('location_id', $location_id, '=');
-      $io->condition('variation_id', $variation_id, '=');
-      $io->execute();
+  /**
+   * Gets the last transaction id for a given location and variation.
+   *
+   * @param $location_id
+   *   Location id.
+   *
+   * @param $variation_id
+   *   Product variation id.
+   *
+   * @return int
+   *   The last location stock transaction id.
+   */
+  public function getLocationStockTransactionLatest($location_id, $variation_id) {
+    $db = \Drupal::database();
+    $query = $db->select('cs_inventory_transaction')
+      ->condition('location_id', $location_id)
+      ->condition('variation_id', $variation_id);
+    $query->addExpression('MAX(trid)', 'max_id');
+    $query->groupBy('location_id');
+    $result = $query
+      ->execute()
+      ->fetch();
+
+    return $result ? $result->max_id : 0;
+  }
+
+  /**
+   * Gets the sum of all stock transactions between a range of transactions.
+   */
+  public function getLocationStockTransactionSum($location_id, $variation_id, $min, $max) {
+    $db = \Drupal::database();
+    $query = $db->select('cs_inventory_transaction', 'it')
+      ->fields('it', ['location_id'])
+      ->condition('location_id', $location_id)
+      ->condition('variation_id', $variation_id)
+      ->condition('trid', $min, '>');
+    if ($max) {
+      $query->condition('trid', $max, '<=');
     }
+    $query->addExpression('SUM(qty)', 'qty');
+    $query->groupBy('location_id');
+    $result = $query->execute()
+      ->fetch();
+
+    return $result ? $result->qty : 0;
   }
 
   /**
    * {@inheritdoc}
    */
-  public function getStockLevel($variation_id, array $locations) {
-    /** @var array $location_info */
-    $location_info = $this->getStockLocationLevel($variation_id, $locations);
-    // Add the quantities together and return.
-    $qty = 0;
+  public function getTotalStockLevel($variation_id, array $locations) {
+    $location_info = $this->getLocationsStockLevels($variation_id, $locations);
+    $total = 0;
     foreach ($location_info as $location) {
-      $qty += $location['qty'] + $location['transactions_qty'];
```

Comment:
The previous code was calculating the real-time current quantity across all stores by combining the computed levels with the as yet un-processed transactions' quantities.  Is this something we really want to be doing?